Should not crash on SonarException: Can not add the same measure twice on org.sonar.api.resources

With sonar-cxx-plugin-0.9.7-SNAPSHOT-RC1.jar, Sonar 5.6.4 and analysis done from TFS Build 
The same project scans OK with sonar-cxx-plugin-0.9.6
It would be better if this error would be reported with better information and no crash occured.
Thanks.

45.8167172Z INFO: Sensor CxxXunitSensor
2017-02-09T13:25:45.8167172Z INFO: Scanner found '1' report files
2017-02-09T13:25:45.8167172Z INFO: Parser will parse '1' report files
2017-02-09T13:25:45.8177173Z INFO: Processing report 'C:\Agent_SLB-8Q06ZZ_work\3\TestResults\GtestReport.xml'
2017-02-09T13:25:45.8177173Z INFO: Parsing 'xUnit' format
2017-02-09T13:25:45.9177273Z INFO: Sensor CxxXunitSensor (done) | time=110ms
2017-02-09T13:25:45.9187274Z INFO: Sensor Unit Test Results Import
2017-02-09T13:25:45.9197275Z INFO: Parsing the Visual Studio Test Results file C:\Agent_SLB-8Q06ZZ_work\3\TestResults\basesrpc_SLB-8Q06ZZ1 2017-02-09 10_57_41.trx
2017-02-09T13:25:49.5600915Z INFO: ------------------------------------------------------------------------
2017-02-09T13:25:49.5710926Z INFO: EXECUTION FAILURE
2017-02-09T13:25:49.5710926Z INFO: ------------------------------------------------------------------------
2017-02-09T13:25:49.5720927Z INFO: Total time: 1:45:16.925s
2017-02-09T13:25:49.6721027Z INFO: Final Memory: 50M/1016M
2017-02-09T13:25:49.6731028Z INFO: ------------------------------------------------------------------------
2017-02-09T13:25:49.6741029Z ##[error]ERROR: Error during SonarQube Scanner execution
2017-02-09T13:25:49.6741029Z ##[error]org.sonar.api.utils.SonarException: Can not add the same measure twice on org.sonar.api.resources.Project@2aea2e4f[id=<null>,key=Cementing_Modeling_QA,qualifier=TRK]: org.sonar.api.measures.Measure@5786d4d1[metricKey=tests,metric=Metric[id=<null>,key=tests,description=<null>,type=INT,direction=0,domain=<null>,name=tests,qualitative=false,userManaged=false,enabled=true,worstValue=<null>,bestValue=<null>,optimizedBestValue=false,hidden=false,deleteHistoricalData=false,decimalScale=<null>],value=3124.0,data=<null>,description=<null>,alertStatus=<null>,alertText=<null>,date=<null>,variation1=<null>,variation2=<null>,variation3=<null>,variation4=<null>,variation5=<null>,url=<null>,personId=<null>,persistenceMode=FULL,fromCore=false]
2017-02-09T13:25:49.6751030Z ##[error]at org.sonar.batch.sensor.DefaultSensorStorage.saveMeasure(DefaultSensorStorage.java:148)
2017-02-09T13:25:49.6751030Z ##[error]at org.sonar.batch.sensor.DefaultSensorStorage.store(DefaultSensorStorage.java:130)
2017-02-09T13:25:49.6751030Z ##[error]at org.sonar.api.batch.sensor.measure.internal.DefaultMeasure.doSave(DefaultMeasure.java:93)
2017-02-09T13:25:49.6761031Z ##[error]at org.sonar.api.batch.sensor.internal.DefaultStorable.save(DefaultStorable.java:43)
2017-02-09T13:25:49.6761031Z ##[error]at org.sonar.plugins.dotnet.tests.UnitTestResultsImportSensor.analyze(UnitTestResultsImportSensor.java:69)
2017-02-09T13:25:49.6771032Z ##[error]at org.sonar.plugins.dotnet.tests.UnitTestResultsImportSensor.execute(UnitTestResultsImportSensor.java:57)
2017-02-09T13:25:49.6771032Z ##[error]at org.sonar.batch.sensor.SensorWrapper.analyse(SensorWrapper.java:57)
2017-02-09T13:25:49.6771032Z ##[error]at org.sonar.batch.phases.SensorsExecutor.executeSensor(SensorsExecutor.java:58)
2017-02-09T13:25:49.6781033Z ##[error]at org.sonar.batch.phases.SensorsExecutor.execute(SensorsExecutor.java:50)
2017-02-09T13:25:49.6781033Z ##[error]at org.sonar.batch.phases.AbstractPhaseExecutor.execute(AbstractPhaseExecutor.java:83)
2017-02-09T13:25:49.6781033Z ##[error]at org.sonar.batch.scan.ModuleScanContainer.doAfterStart(ModuleScanContainer.java:192)
2017-02-09T13:25:49.6791034Z ##[error]at org.sonar.core.platform.ComponentContainer.startComponents(ComponentContainer.java:142)
2017-02-09T13:25:49.6791034Z ##[error]at org.sonar.core.platform.ComponentContainer.execute(ComponentContainer.java:127)
2017-02-09T13:25:49.6791034Z ##[error]at org.sonar.batch.scan.ProjectScanContainer.scan(ProjectScanContainer.java:241)
2017-02-09T13:25:49.6801035Z ##[error]at org.sonar.batch.scan.ProjectScanContainer.scanRecursively(ProjectScanContainer.java:236)
2017-02-09T13:25:49.6801035Z ##[error]at org.sonar.batch.scan.ProjectScanContainer.doAfterStart(ProjectScanContainer.java:226)
2017-02-09T13:25:49.6811036Z ##[error]at org.sonar.core.platform.ComponentContainer.startComponents(ComponentContainer.java:142)
2017-02-09T13:25:49.6811036Z ##[error]at org.sonar.core.platform.ComponentContainer.execute(ComponentContainer.java:127)
2017-02-09T13:25:49.6811036Z ##[error]at org.sonar.batch.task.ScanTask.execute(ScanTask.java:47)
2017-02-09T13:25:49.6821037Z ##[error]at org.sonar.batch.task.TaskContainer.doAfterStart(TaskContainer.java:86)
2017-02-09T13:25:49.6821037Z ##[error]at org.sonar.core.platform.ComponentContainer.startComponents(ComponentContainer.java:142)
2017-02-09T13:25:49.6821037Z ##[error]at org.sonar.core.platform.ComponentContainer.execute(ComponentContainer.java:127)
2017-02-09T13:25:49.6831038Z ##[error]at org.sonar.batch.bootstrap.GlobalContainer.executeTask(GlobalContainer.java:106)
2017-02-09T13:25:49.6831038Z ##[error]at org.sonar.batch.bootstrapper.Batch.executeTask(Batch.java:119)
2017-02-09T13:25:49.6831038Z ##[error]at org.sonarsource.scanner.api.internal.batch.BatchIsolatedLauncher.execute(BatchIsolatedLauncher.java:62)
2017-02-09T13:25:49.6841039Z ##[error]at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
2017-02-09T13:25:49.6841039Z ##[error]at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
2017-02-09T13:25:49.6851040Z ##[error]at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
2017-02-09T13:25:49.6851040Z ##[error]at java.lang.reflect.Method.invoke(Method.java:498)
2017-02-09T13:25:49.6851040Z ##[error]at org.sonarsource.scanner.api.internal.IsolatedLauncherProxy.invoke(IsolatedLauncherProxy.java:60)
2017-02-09T13:25:49.6861041Z ##[error]at com.sun.proxy.$Proxy0.execute(Unknown Source)
2017-02-09T13:25:49.6861041Z ##[error]at org.sonarsource.scanner.api.EmbeddedScanner.doExecute(EmbeddedScanner.java:233)
2017-02-09T13:25:49.6861041Z ##[error]at org.sonarsource.scanner.api.EmbeddedScanner.runAnalysis(EmbeddedScanner.java:151)
2017-02-09T13:25:49.6871042Z ##[error]at org.sonarsource.scanner.cli.Main.runAnalysis(Main.java:110)
2017-02-09T13:25:49.6871042Z ##[error]at org.sonarsource.scanner.cli.Main.execute(Main.java:74)
2017-02-09T13:25:49.6871042Z ##[error]at org.sonarsource.scanner.cli.Main.main(Main.java:61)
2017-02-09T13:25:49.6881043Z ##[error]ERROR:
2017-02-09T13:25:49.6881043Z ##[error]ERROR: Re-run SonarQube Scanner using the -X switch to enable full debug logging.
2017-02-09T13:25:49.8881243Z ##[error]The SonarQube Scanner did not complete successfully
2017-02-09T13:25:49.8891244Z 14:25:49.779  Creating a summary markdown file...
2017-02-09T13:25:49.8891244Z ##[error]14:25:49.784  Post-processing failed. Exit code: 1
2017-02-09T13:25:49.8941249Z ##[debug]The build was not triggered by a Pull Request, not processing code analysis comments
2017-02-09T13:25:49.9011256Z ##[debug]Variable read: MSBuild.SonarQube.Internal.IncludeFullReport = true

--- a/sonar-cxx-plugin/src/main/java/org/sonar/plugins/cxx/tests/dotnet/CxxUnitTestResultsProvider.java
+++ b/sonar-cxx-plugin/src/main/java/org/sonar/plugins/cxx/tests/dotnet/CxxUnitTestResultsProvider.java
@@ -20,7 +20,11 @@
 package org.sonar.plugins.cxx.tests.dotnet;
 
 import org.sonar.api.batch.bootstrap.ProjectDefinition;
+import org.sonar.api.batch.sensor.SensorContext;
 import org.sonar.api.config.Settings;
+import org.sonar.api.utils.log.Logger;
+import org.sonar.api.utils.log.Loggers;
+import org.sonar.plugins.cxx.utils.CxxUtils;
 import org.sonar.plugins.dotnet.tests.UnitTestConfiguration;
 import org.sonar.plugins.dotnet.tests.UnitTestResultsAggregator;
 import org.sonar.plugins.dotnet.tests.UnitTestResultsImportSensor;
@@ -30,27 +34,44 @@ public class CxxUnitTestResultsProvider {
   public static final String VISUAL_STUDIO_TEST_RESULTS_PROPERTY_KEY = "sonar.cxx.vstest.reportsPaths";
   public static final String NUNIT_TEST_RESULTS_PROPERTY_KEY = "sonar.cxx.nunit.reportsPaths";
   public static final String XUNIT_TEST_RESULTS_PROPERTY_KEY = "sonar.cxx.xunit.reportsPaths";
-  
+
   private static final UnitTestConfiguration UNIT_TEST_CONF = new UnitTestConfiguration(
-          VISUAL_STUDIO_TEST_RESULTS_PROPERTY_KEY,
-          NUNIT_TEST_RESULTS_PROPERTY_KEY,
-          XUNIT_TEST_RESULTS_PROPERTY_KEY);
-  
+    VISUAL_STUDIO_TEST_RESULTS_PROPERTY_KEY,
+    NUNIT_TEST_RESULTS_PROPERTY_KEY,
+    XUNIT_TEST_RESULTS_PROPERTY_KEY);
+
   private CxxUnitTestResultsProvider() {
   }
 
   public static class CxxUnitTestResultsAggregator extends UnitTestResultsAggregator {
-    
+
     public CxxUnitTestResultsAggregator(Settings settings) {
       super(UNIT_TEST_CONF, settings);
     }
   }
 
   public static class CxxUnitTestResultsImportSensor extends UnitTestResultsImportSensor {
-    
+
+    public static final Logger LOG = Loggers.get(CxxUnitTestResultsImportSensor.class);
+
     public CxxUnitTestResultsImportSensor(CxxUnitTestResultsAggregator unitTestResultsAggregator, ProjectDefinition projectDef) {
       super(unitTestResultsAggregator, projectDef);
     }
 
+    @Override
+    public void execute(SensorContext context) {
+      try {
+        super.execute(context);
+      } catch (Exception e) {
+        String msg = new StringBuilder()
+          .append("Cannot feed the data into SonarQube, details: '")
+          .append(e)
+          .append("'")
+          .toString();
+        LOG.error(msg);
+        CxxUtils.validateRecovery(e, context.settings());
+      }
+    }
   }
+
 }

